fblock.go

```diff
@@ -0,0 +1,455 @@
+	ExpansionSize  int64 `json:"expansionsize"` // Stored as a varint
+	ExpansionBytes Bytes `json:"expansiondata"`
+
+	// Number of txs in the block
+	TransactionCount uint32 `json:"txcount"`
+	// Number of bytes contained in the body
+	BodySize uint32 `json:"bodysize"`
+}
+
+// IsPopulated returns true if fb has already been successfully populated by a
+// call to Get. IsPopulated returns false if fb.Transactions is nil.
+func (fb FBlock) IsPopulated() bool {
+	return len(fb.Transactions) > 0 && // FBlocks always contain at least a coinbase
+		fb.Header.BodyMR != nil &&
+		fb.Header.PrevKeyMR != nil && // TODO: Why not check the computed keymr?
+		fb.Header.PrevLedgerKeyMR != nil
+}
+
+// Get queries factomd for the Factoid Block at fb.Header.Height or fb.KeyMR.
+// After a successful call, the Transactions will all be populated.
+func (fb *FBlock) Get(c *Client) (err error) {
+	if fb.IsPopulated() {
+		return nil
+	}
+
+	if fb.KeyMR != nil {
+		orig := fb.KeyMR
+		defer func() { // Ensure we got the block we asked for
+			if err != nil {
+				return
+			}
+
+			if fb.KeyMR != nil && *orig != *fb.KeyMR {
+				err = fmt.Errorf("invalid key merkle root")
+				return
+			}
+		}()
+
+		params := struct {
+			Hash *Bytes32 `json:"hash"`
+		}{Hash: fb.KeyMR}
+		var result struct {
+			Data Bytes `json:"data"`
+		}
+		if err := c.FactomdRequest("raw-data", params, &result); err != nil {
+			return err
+		}
+		return fb.UnmarshalBinary(result.Data)
+	}
+
+	params := struct {
+		Height uint32 `json:"height"`
+	}{fb.Header.Height}
+	result := struct {
+		// We will ignore all the other fields, and just unmarshal from the raw.
+		RawData Bytes `json:"rawdata"`
+	}{}
+	if err := c.FactomdRequest("fblock-by-height", params, &result); err != nil {
+		return err
+	}
+
+	return fb.UnmarshalBinary(result.RawData)
+}
+
+const (
+	FBlockMinHeaderLen = 32 + // Factoid ChainID
+		32 + // BodyMR
+		32 + // PrevKeyMR
+		32 + // PrevLedgerKeyMR
+		8 + // EC Exchange Rate
+		4 + // DB Height
+		0 + // Header Expansion size (varint)
+		0 + // Header Expansion Area (Min 0)
+		4 + // Transaction Count
+		4 // Body Size
+
+	FBlockMinTotalLen = FBlockMinHeaderLen
+)
+
+// UnmarshalBinary unmarshals raw directory block data.
+//
+// Header
+// [Factoid Block ChainID (Bytes32{31:0x0a})] +
+// [BodyMR (Bytes32)] +
+// [PrevKeyMR (Bytes32)] +
+// [PrevLedgerKeyMR (Bytes32)] +
+// [Exchange Rate (8 bytes)] +
+// [DB Height (4 bytes)] +
+// [Header Expansion size (Bytes)] +
+// [Header Expansion Area (Bytes)] +
+// [Transaction Count (4 bytes)] +
+// [Body Size (4 bytes)] +
+//
+// Body
+// [Tx 0 (Bytes)] +
+// ... +
+// [Tx N (Bytes)] +
+//
+// https://github.com/FactomProject/FactomDocs/blob/master/factomDataStructureDetails.md#factoid-block
+func (fb *FBlock) UnmarshalBinary(data []byte) (err error) {
+	// Because the length of an FBlock is hard to define up front, we will catch
+	// any sort of out of bound errors in a recover
+	defer func() {
```

Comment:
I agree I do not like using a recover(), but there is a lot of unknown lengths here. An fblock contains all txs in their entirety. Each tx has a variable number of inputs/outputs and varints for the amounts. I will probably go through and try to catch more, but there is so many edge cases, not just out of bounds panics.

E.g: Someone could send you data saying there is maxint number of inputs, and you'd try to allocate a factoid tx slice outside the limits of your system memory. 

I do want to add more protections, but this was my catch all, as I'd need to think of all the possible ways to crash this. And it's more than the other structures.

Reply:
If we run out of memory there is nothing we can do and that is the exact situation where an unrecovered panic is appropriate. 

This recover hides programming or deeper errors and reduces them to a single error with no context.

Reply:
I am saying if they try to unmarshal a blob of data is that totally malformed, currently it will panic.

As a developer, I assume if I pass a bad fblock data blob into this function, it will give me an error if there is a problem with the data. I don't expect panics from libraries, as panics are not like exceptions.

The OOM panic is because the binary data is wrong, and it says to allocated some massive number of transactions, which obviously fails. We can put safety checks at places, but we'd be choosing arbitrary magic numbers for the maximum number of txs in a block for example

Reply:
I did as much as I could to catch things and dropped the defer.